sklearn/utils/_array_api.py

```diff
@@ -555,6 +555,9 @@ def _fill_or_add_to_diagonal(array, value, xp, add_value=True, wrap=False):
         array_flat[:end:step] += value
     else:
         array_flat[:end:step] = value
+    # `array_flat` is not always a view on `array` (e.g., certain array types
+    # created via parallel processing), thus we need to return reshaped `array_flat`.
```

Comment:
Took me a while to realise this was happening.

Interestingly `array_api_strict-device1-float64` fails but `array_api_strict-device2-float32` fails.
Note that I did try using `reshape(copy=False)` but this fails on array_api_strict with this error:

```python
    if copy is False and not np.shares_memory(data, reshaped):
        raise AttributeError("Incompatible shape for in-place modification.")
```

and failed on pytorch with a not implemented error.

I don't fully get combination of array type is causing this and how the parallel plays into it so please advise if I can make this comment more informative/correct?

Reply:
Amended comment to:
```python
    # `array_flat` is not always a view on `array` (e.g. for certain array types that
    # were filled via parallel processing), thus we need to return reshaped
    # `array_flat`.
```

Reply:
Maybe you could actually give an example of such parallel code in the comment to avoid having the reader try to look for it?

Reply:
Yes good point.
I still don't know why this only happens with certain array types though, but I am thinking the parallel plays a part.